src/charm.py

```diff
@@ -0,0 +1,104 @@
+#!/usr/bin/env python3
+
+import logging
+
+from ops.main import main
+from ops.pebble import Layer
+from ops.charm import CharmBase
+from ops.model import ActiveStatus, WaitingStatus, MaintenanceStatus
+
+logger = logging.getLogger(__name__)
+
+
+class TrainingOperatorCharm(CharmBase):
+    """A Juju Charm for Training Operator"""
+
+    def __init__(self, *args):
+        super().__init__(*args)
+
+        self.logger = logging.getLogger(__name__)
+
+        self._name = self.model.app.name
+        self._namespace = self.model.name
+        self._manager_service = "manager"
+
+        self.framework.observe(self.on.install, self._on_config_changed)
+        self.framework.observe(self.on.config_changed, self._on_config_changed)
+        self.framework.observe(
+            self.on.training_operator_pebble_ready,
+            self._on_training_operator_pebble_ready,
+        )
+
+    @property
+    def _training_operator_layer(self) -> Layer:
+        """Property which returns a pre-configured Pebble layer.
+
+        The Pbble layer contains information about how to start the manager service,
+        and sets the relevant environment variables for the process to start successfully.
+
+        Returns:
```

Comment:
+1 on type annotations (and even using mypy), the alertmanager-charm implements this already.

That said, we do generally encourage Google Style docstrings both in the Operator Framework, and in charms, and this is indeed a feature of google style docstrings.

Reply:
+6 to type annotations.  I wish the juju libraries used them - it would make learning the API way easier

On the level of completeness of our docstrings, I'm torn.  I agree with @knkski that `Returns: Layer: a Pebble layer specification for the manager service` feels very redundant when they're type hinting (unless there's some non-obvious quality in the returned object, like `a Pebble layer with X`).  But if the docstring is used for some other documentation (eg mkdocs), they might be the only place that info gets presented.

I lean slightly toward including them but making them easy to create/maintain.  Not sure if IDEs automate typing for their auto-google-docstrings, but I've seen examples of putting it in [commit hooks](https://towardsdatascience.com/python-type-hints-docstrings-7ec7f6d3416b).  Not a solution you can push on others, but maybe something that balances the effort/reward @knkski?

Reply:
TBH, I would write a docstring of simply `"""Returns a pre-configured Pebble layer."""`. There's nothing surprising going on in this method, so there's no reason to comment further.

Conveniently, I just looked this up in the Google style guide, which says this:

> [Returns:] may also be omitted if the docstring starts with Returns or Yields (e.g. """Returns row from Bigtable as a tuple of strings.""") and the opening sentence is sufficient to describe the return value.